fix: shrink queued bodies and headers when drained or split

Rjected opened the pull request:
This fixes an OOM in the bodies stage and reduces overall memory allocations in the downloaders, by shrinking vectors whenever we `split` or `drain` them. Previously the headers downloader would effectively clone `queued_validated_headers` in terms of allocations, because `split_off` does not change the original vec's capacity. Because `queued_validated_headers` could be on the order of gigabytes, this caused an OOM.

Fixes #3175

gakonst reviewed (approved):
supportive - any way of testing this?

mattsse commented on crates/net/downloaders/src/headers/reverse_headers.rs:
```diff
@@ -333,6 +333,7 @@ where
                     .count();
                 // removes all headers that are higher than current target
                 self.queued_validated_headers.drain(..skip);
+                self.queued_validated_headers.shrink_to_fit();
```
I don't think this is necessary here, because this merely removes a few headers from the front, and we will go through `next_batch` eventually

mattsse commented on crates/net/downloaders/src/headers/reverse_headers.rs:
```diff
@@ -616,6 +617,21 @@ where
+        // To prevent these allocations from leaking to the consumer, we shrink the capacity of the
+        // two buffers. The total memory allocated should then be not much more than the original
+        // size of `queued_validated_headers`.
+        rem.shrink_to_fit();
```
hmm, I think this is redundant because `split_off` only allocates required capacity

mattsse reviewed (changes requested):
very nice find.

This also applies to the BianryHeaps which are just Vec wrappers, so we need to shrink them as well.

I think we can get away with having our own shrink_to fit, where we delegate the calls and only call this once when we return the next batch?

Rjected commented:
> supportive as discussed - any way of testing this?

Just added some checks in the existing tests that make sure `len == capacity` for downloaded values, when the stream batch size is lower than the number of values that should be produced.

also added another test for one-by-one download with a larger request limit, so `queued_validated_headers` is hopefully larger than the batch size in that test, which should trigger the necessary condition.

rkrasiuk reviewed (approved):
majestic!

Rjected commented:
@mattsse added calls to `shrink_to_fit` whenever buffered responses are popped in 781d786e8d2f2de9b9e97e0e3748482d1815ec1e

mattsse commented on crates/net/downloaders/src/headers/reverse_headers.rs:
```diff
@@ -557,6 +557,9 @@ where
                 Ordering::Equal => {
                     let OrderedHeadersResponse { headers, request, peer_id } =
                         PeekMut::pop(next_response);
+
+                    // shrink the buffer so that it doesn't grow indefinitely
+                    self.buffered_responses.shrink_to_fit();
```
having this inside a loop is not ideal,

I think we can move it below the loop, since the error case would retrigger a new download

mattsse reviewed (changes requested):
smol nits

just to be safe I'd also like to replace all `.clear()` calls in the `fn clear()` functions of bodies and headers downloader with new initializations because clear does not change capacity

Rjected commented on crates/net/downloaders/src/bodies/bodies.rs:
```diff
@@ -222,6 +222,8 @@ where
     fn pop_buffered_response(&mut self) -> Option<OrderedBodiesResponse> {
         let resp = self.buffered_responses.pop()?;
+        // shrink the buffer so that it doesn't grow indefinitely
+        self.buffered_responses.shrink_to_fit();
```
ah right, these should definitely be moved outside loops

sigridjineth commented:
thks for resolving this.
